rust/connlib/tunnel/src/tests.rs

```diff
@@ -74,103 +108,659 @@ impl StateMachineTest for TunnelTest {
+            any_advanced = true;
+
+            self.client
+                .span
+                .in_scope(|| self.client.state.handle_timeout(now));
+        };
+
+        if self.gateway.state.poll_timeout().is_some_and(|t| t <= now) {
+            any_advanced = true;
+
+            self.gateway
+                .span
+                .in_scope(|| self.gateway.state.handle_timeout(now, utc_now))
+        };
 
-        // We select which resource to send to based on the _longest match_ of the IP network.
-        // We may have resources with overlapping IP ranges so it is important that we do this the same way as connlib.
-        let Some((_, resource)) = self.client_resources.longest_match(dst) else {
+        if self.relay.state.poll_timeout().is_some_and(|t| t <= now) {
+            any_advanced = true;
+
+            self.relay
+                .span
+                .in_scope(|| self.relay.state.handle_timeout(now))
+        };
+
+        any_advanced
+    }
+
+    fn send_icmp_packet_client_to_gateway(
+        &mut self,
+        src: impl Into<IpAddr>,
+        dst: impl Into<IpAddr>,
+    ) -> Option<(Transmit<'static>, Option<SocketAddr>)> {
+        let transmit = self.client.span.in_scope(|| {
+            self.client.state.encapsulate(
+                ip_packet::make::icmp_request_packet(src.into(), dst.into()),
+                self.now,
+            )
+        })?;
+        let transmit = transmit.into_owned();
+        let sending_socket = self.client.sending_socket_for(transmit.dst);
+
+        Some((transmit, sending_socket))
+    }
+
+    /// Dispatches a [`Transmit`] to the correct component.
+    ///
+    /// This function is basically the "network layer" of our tests.
+    /// It takes a [`Transmit`] and checks, which component accepts it, i.e. has configured the correct IP address.
+    /// Our tests don't have a concept of a network topology.
+    /// This means, components can have IP addresses in completely different subnets, yet this function will still "route" them correctly.
+    fn dispatch_transmit(
+        &mut self,
+        transmit: Transmit,
+        sending_socket: Option<SocketAddr>,
+        buffered_transmits: &mut VecDeque<(Transmit<'static>, Option<SocketAddr>)>,
+    ) {
+        let dst = transmit.dst;
+        let payload = &transmit.payload;
+
+        let Some(src) = sending_socket else {
+            tracing::warn!("Dropping packet to {dst}: no socket");
             return;
         };
-        let resource = resource.id;
 
-        // Only expect a new connection intent if the last one was older than 2s
-        if !self.will_send_new_connection_intent(&resource) {
+        if self.relay.wants(dst) {
+            self.relay
+                .handle_packet(payload, src, dst, self.now, buffered_transmits);
+
+            return;
+        }
+
+        let src = transmit
+            .src
+            .expect("to have handled all packets without src via relays");
```

Comment:
https://doc.rust-lang.org/std/error/index.html#common-message-styles
```suggestion
            .expect("all packets without src should have been handled via relays");
```